diff --git a/base/values/struct_value.h b/base/values/struct_value.h
index e7237f62..9dd58e2e 100644
--- a/base/values/struct_value.h
+++ b/base/values/struct_value.h
@@ -367,9 +367,9 @@ class AbstractStructValue : public StructValue,
       ValueFactory& value_factory, int64_t number) const = 0;
 
   virtual absl::StatusOr<QualifyResult> Qualify(
-      ValueFactory& value_factory,
-      absl::Span<const SelectQualifier> select_qualifiers,
-      bool presence_test) const {
+      ValueFactory& value_factory ABSL_ATTRIBUTE_UNUSED,
+      absl::Span<const SelectQualifier> select_qualifiers ABSL_ATTRIBUTE_UNUSED,
+      bool presence_test ABSL_ATTRIBUTE_UNUSED) const {
     return absl::UnimplementedError("Qualify not supported.");
   }
 
diff --git a/eval/public/cel_expr_builder_factory.h b/eval/public/cel_expr_builder_factory.h
index 7321e29a..a7e4fc42 100644
--- a/eval/public/cel_expr_builder_factory.h
+++ b/eval/public/cel_expr_builder_factory.h
@@ -1,6 +1,7 @@
 #ifndef THIRD_PARTY_CEL_CPP_EVAL_PUBLIC_CEL_EXPR_BUILDER_FACTORY_H_
 #define THIRD_PARTY_CEL_CPP_EVAL_PUBLIC_CEL_EXPR_BUILDER_FACTORY_H_
 
+#include "absl/log/absl_log.h"
 #include "google/protobuf/descriptor.h"
 #include "eval/public/cel_expression.h"
 #include "eval/public/cel_options.h"
diff --git a/eval/public/cel_value.h b/eval/public/cel_value.h
index 6ad5715b..40026538 100644
--- a/eval/public/cel_value.h
+++ b/eval/public/cel_value.h
@@ -25,7 +25,6 @@
 #include "absl/base/attributes.h"
 #include "absl/base/macros.h"
 #include "absl/base/optimization.h"
-#include "absl/log/absl_log.h"
 #include "absl/status/status.h"
 #include "absl/status/statusor.h"
 #include "absl/strings/str_cat.h"
@@ -488,9 +487,8 @@ class CelValue {
   explicit CelValue(T value) : value_(value) {}
 
   // Crashes with a null pointer error.
-  static void CrashNullPointer(Type type) ABSL_ATTRIBUTE_COLD {
-    ABSL_LOG(FATAL) << "Null pointer supplied for "
-                    << TypeName(type);  // Crash ok
+  static void CrashNullPointer(Type) ABSL_ATTRIBUTE_COLD {
+    ABSL_ASSERT(false);
   }
 
   // Null pointer checker for pointer-based types.
@@ -501,11 +499,9 @@ class CelValue {
   }
 
   // Crashes with a type mismatch error.
-  static void CrashTypeMismatch(Type requested_type,
-                                Type actual_type) ABSL_ATTRIBUTE_COLD {
-    ABSL_LOG(FATAL) << "Type mismatch"                             // Crash ok
-                    << ": expected " << TypeName(requested_type)   // Crash ok
-                    << ", encountered " << TypeName(actual_type);  // Crash ok
+  static void CrashTypeMismatch(Type,
+                                Type) ABSL_ATTRIBUTE_COLD {
+    ABSL_ASSERT(false);
   }
 
   // Gets value of type specified
diff --git a/eval/public/structs/BUILD b/eval/public/structs/BUILD
index bbfd8b22..13d9c09b 100644
--- a/eval/public/structs/BUILD
+++ b/eval/public/structs/BUILD
@@ -194,7 +194,6 @@ cc_library(
     deps = [
         ":legacy_any_packing",
         ":legacy_type_adapter",
-        "//base:type",
         "@com_google_absl//absl/types:optional",
     ],
 )
diff --git a/eval/public/structs/field_access_impl.cc b/eval/public/structs/field_access_impl.cc
index 952fcebc..88d42256 100644
--- a/eval/public/structs/field_access_impl.cc
+++ b/eval/public/structs/field_access_impl.cc
@@ -25,6 +25,7 @@
 #include "google/protobuf/arena.h"
 #include "google/protobuf/map_field.h"
 #include "absl/container/flat_hash_set.h"
+#include "absl/log/absl_log.h"
 #include "absl/status/status.h"
 #include "absl/status/statusor.h"
 #include "absl/strings/str_cat.h"
diff --git a/eval/public/structs/legacy_type_adapter.h b/eval/public/structs/legacy_type_adapter.h
index 593dbf5c..51226b45 100644
--- a/eval/public/structs/legacy_type_adapter.h
+++ b/eval/public/structs/legacy_type_adapter.h
@@ -65,9 +65,10 @@ class LegacyTypeMutationApis {
       CelValue::MessageWrapper::Builder& instance) const = 0;
 
   virtual absl::Status SetFieldByNumber(
-      int64_t field_number, const CelValue& value,
-      cel::MemoryManagerRef memory_manager,
-      CelValue::MessageWrapper::Builder& instance) const {
+      int64_t field_number ABSL_ATTRIBUTE_UNUSED,
+      const CelValue& value ABSL_ATTRIBUTE_UNUSED,
+      cel::MemoryManagerRef memory_manager ABSL_ATTRIBUTE_UNUSED,
+      CelValue::MessageWrapper::Builder& instance ABSL_ATTRIBUTE_UNUSED) const {
     return absl::UnimplementedError("SetFieldByNumber is not yet implemented");
   }
 };
@@ -116,8 +117,9 @@ class LegacyTypeAccessApis {
   //   whether the leaf field is set to a non-default value.
   virtual absl::StatusOr<LegacyQualifyResult> Qualify(
       absl::Span<const cel::SelectQualifier>,
-      const CelValue::MessageWrapper& instance, bool presence_test,
-      cel::MemoryManagerRef memory_manager) const {
+      const CelValue::MessageWrapper& instance ABSL_ATTRIBUTE_UNUSED,
+      bool presence_test ABSL_ATTRIBUTE_UNUSED,
+      cel::MemoryManagerRef memory_manager ABSL_ATTRIBUTE_UNUSED) const {
     return absl::UnimplementedError("Qualify unsupported.");
   }
 
diff --git a/eval/public/structs/legacy_type_provider.h b/eval/public/structs/legacy_type_provider.h
index eea5d44b..255c6259 100644
--- a/eval/public/structs/legacy_type_provider.h
+++ b/eval/public/structs/legacy_type_provider.h
@@ -16,7 +16,6 @@
 #define THIRD_PARTY_CEL_CPP_EVAL_PUBLIC_STRUCTS_TYPE_PROVIDER_H_
 
 #include "absl/types/optional.h"
-#include "base/type_provider.h"
 #include "eval/public/structs/legacy_any_packing.h"
 #include "eval/public/structs/legacy_type_adapter.h"
 
@@ -26,8 +25,10 @@ namespace google::api::expr::runtime {
 //
 // Note: This API is not finalized. Consult the CEL team before introducing new
 // implementations.
-class LegacyTypeProvider : public cel::TypeProvider {
+class LegacyTypeProvider {
  public:
+  virtual ~LegacyTypeProvider() = default;
+
   // Return LegacyTypeAdapter for the fully qualified type name if available.
   //
   // nullopt values are interpreted as not present.
diff --git a/internal/copy_on_write.h b/internal/copy_on_write.h
index 654f2aae..4c9556bb 100644
--- a/internal/copy_on_write.h
+++ b/internal/copy_on_write.h
@@ -48,7 +48,7 @@ class ABSL_ATTRIBUTE_TRIVIAL_ABI CopyOnWrite final {
 
     template <typename... Args,
               typename = std::enable_if_t<std::is_constructible_v<T, Args...>>>
-    explicit Rep(Args&&... args) : value(std::forward<Args>(value)...) {}
+    explicit Rep(Args&&... args) : value(std::forward<Args>(args)...) {}
 
     Rep(const Rep&) = delete;
     Rep(Rep&&) = delete;
